tests/backend/asgi/authenticated/utils.py

```diff
@@ -0,0 +1,116 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPL-3.0 2016-present Scille SAS
+
+from __future__ import annotations
+from base64 import b64encode
+from structlog import get_logger
+from typing import Callable, Optional, Tuple
+from quart import Response
+from quart.datastructures import Headers
+from quart.typing import TestClientProtocol
+
+from parsec.crypto import SigningKey
+from parsec.api.protocol.types import DeviceID, OrganizationID
+from parsec.api.version import API_VERSION, ApiVersion
+from parsec.api.protocol.ping import ping_serializer
+from parsec._parsec import DateTime
+
+
+logger = get_logger()
+
+
+def dummy_request_with_reply() -> Tuple[bytes, Callable[[bytes], dict], dict]:
+    """
+    Create a dummy request also provide the following:
+
+    - a parser to parse the receive reply from the server
+    - its expected reply
+    """
+    request_payload = "hello world"
+
+    req = ping_serializer.req_dumps({"cmd": "ping", "ping": request_payload})
+
+    def reply_parser(raw_data) -> dict:
+        return ping_serializer.rep_loads(raw_data)
+
+    rep = {"status": "ok", "pong": request_payload}
+
+    return (req, reply_parser, rep)
+
+
+def signing_request(
+    signing_key: SigningKey, device_id: DeviceID, timestamp: DateTime, body: bytes
+) -> str:
+    """
+    Return the signature to validate a request
+    """
+    formatted_datetime = timestamp.to_rfc3339_string()
+    data_to_sign = b64encode(str(device_id).encode()) + formatted_datetime.encode() + body
+    signed_message = signing_key.sign_only_signature(data_to_sign)
+    signed_message_b64 = b64encode(signed_message).decode()
+    logger.debug(
+        f"signing request for device {device_id}",
+        timestamp={formatted_datetime},
+        signature={signed_message_b64},
+    )
+    return signed_message_b64
+
+
+async def send_authenticated_request(
+    client: TestClientProtocol, org_id: OrganizationID, headers: Headers, body: bytes
+) -> Response:
+    return await client.post(f"/authenticated/{org_id}", headers=headers, data=body)
+
+
+class AuthenticatedHeadersBuilder:
+    timestamp: Optional[DateTime] = None
+    signature: Optional[bytes] = None
+    author: Optional[DeviceID] = None
+
+    def __init__(
+        self,
+        api_version: Optional[ApiVersion] = API_VERSION,
+        content_type: Optional[str] = "application/msgpack",
+        authorization: Optional[str] = "PARSEC-SIGN-ED25519",
+    ) -> None:
```

Comment:
It's unclear why some attributes are set with the regular constructor and why others are set with the builder pattern.

Why not simply have a builder function?
```python
def build_authenticated_headers(
    api_version: Optional[ApiVersion] = API_VERSION,
    content_type: Optional[str] = "application/msgpack",
    authorization: Optional[str] = "PARSEC-SIGN-ED25519",
    timestamp: Optional[DateTime] = None,
    signature: Optional[bytes] = None,
    author: Optional[DeviceID] = None,
) -> Headers:
 [...]

Reply:
done